File: gtci/00-warmup/cpp/2-reverse-vowels.cpp
#include <cctype>
#include <iostream>
#include <string>
#include <unordered_set>

class Solution {
   public:
    std::string reverseVowels(std::string s) {
        const std::unordered_set<char> vowels{'a', 'e', 'i', 'o', 'u'};

        std::string::iterator left = s.begin();
        std::string::iterator right = s.end() - 1;

        while (left < right) {
            if (vowels.find(tolower(*left)) == vowels.end()) left += 1;

            if (vowels.find(tolower(*right)) == vowels.end()) right -= 1;

            if (vowels.find(tolower(*left)) != vowels.end() &&
                vowels.find(tolower(*right)) != vowels.end()) {
                std::swap(*left, *right);

                left += 1;
                right -= 1;
            }
        }

        return s;
    }
};

#include <cassert>

int main() {
    Solution sol;
    assert(sol.reverseVowels("hello") == "holle");
    assert(sol.reverseVowels("AEIOU") == "UOIEA");
    assert(sol.reverseVowels("DesignGUrus") == "DusUgnGires");
    std::cout << "All test cases passed." << std::endl;
    return 0;
}

File: gtci/00-warmup/python3/2-reverse-vowels.py
def reverseVowels(self, s: str) -> str:
    s_list = list(s)
    vowels = {"a", "e", "i", "o", "u"}
    left = 0
    right = len(s) - 1

    while left < right:
        if s[left].lower() not in vowels:
            left += 1

        if s[right].lower() not in vowels:
            right -= 1

        if s[left].lower() in vowels and s[right].lower() in vowels:
            s_list[left], s_list[right] = s_list[right], s_list[left]
            left += 1
            right -= 1

    return "".join(s_list)


if __name__ == "__main__":
    print(reverseVowels(None, "hello"))  # holle
    print(reverseVowels(None, "AEOIU"))  # UOIEA
    print(reverseVowels(None, "DesignGUrus"))  # DusUgnGires
